lib/tpm2_util.c

```diff
@@ -1100,11 +1102,142 @@ TPMI_ALG_HASH *param_hash_algorithm, tpm2_session **session) {
+        SESSION_PREFIX_LEN)) {
+            sessions_auths_or_paths[session_idx] += SESSION_PREFIX_LEN;
+        }
+    }
+
+    if (!input_session_count) {
+        return false;
+    }
+
+    return true;
+}
+
+static TPMI_ALG_HASH calc_cphash_alg_from_session_types(ESYS_CONTEXT *ectx,
+const char **adjusted_session_paths) {
+
+    TPMI_ALG_HASH rethash = TPM2_ALG_SHA256;
+    size_t session_idx = 0;
+    for (session_idx = 0; session_idx < MAX_SESSION_CNT; session_idx++) {
+        if (!adjusted_session_paths[session_idx]) {
+            continue;
+        }
+
+        tpm2_session *session = NULL;
+        tool_rc rc = tpm2_session_restore(ectx,
+        adjusted_session_paths[session_idx], false, &session);
+        if (rc != tool_rc_success) {
+            LOG_WARN("Could not restore session #%s",
+            adjusted_session_paths[session_idx]);
+            tpm2_session_close(&session);
+            continue;
+        }
+
+        TPM2_SE session_type = tpm2_session_get_type(session);
+        if (session_type != TPM2_SE_HMAC && session_type != TPM2_SE_POLICY) {
+            tpm2_session_close(&session);
+            continue;
+        }
+
+        rethash = tpm2_session_get_authhash(session);
+
+        TPMA_SESSION attrs = 0;
+        ESYS_TR session_handle = tpm2_session_get_handle(session);
+        tool_rc tmp_rc = tpm2_sess_get_attributes(ectx, session_handle, &attrs);
+        UNUSED(tmp_rc);
+        if (attrs & TPMA_SESSION_AUDIT) {
+            tpm2_session_close(&session);
+            break;
+        }
+    }
+
+    return rethash;
+}
+
+TPMI_ALG_HASH calc_cphash_alg_from_cphash_path(const char *cphash_path) {
+
+    if (!cphash_path) {
+        return TPM2_ALG_ERROR;
+    }
+
     /*
-     * Parameter hash alg is setup as that of the session alg of first session
+     * Expecting single token, so tokenize just once.
      */
-    if (param_hash_algorithm) {
-        *param_hash_algorithm = tpm2_session_get_authhash(session[0]);
+    char *str = (char *)cphash_path;
+    char *token = strtok(str, ":");
+    if (!token) {
+        return TPM2_ALG_ERROR;
     }
-    return tool_rc_success;
+
+    TPMI_ALG_HASH hashalg = tpm2_alg_util_from_optarg(token,
+    tpm2_alg_util_flags_hash);
+    if (hashalg == TPM2_ALG_ERROR) {
+        LOG_WARN("Invalid/ Unspecified policy digest algorithm.");
+    }
+
+    return hashalg;
 }
+
+/*
+ * It should be noted that the auths aren't checked when calculating the cpHash,
+ * instead the sessions are consumed to determine the cpHash algorithm.
+ *
+ * 1. If cphash_path is preceded with <halg>: use that as cphash-halg return
+ * Otherwise
+ *
+ *  Consume session only if it is a policy-session or an hmac-session
+ *  1. If only hmac or policy session is specified, return that session's halg
+ *  2. If hmac-session with audit is specified, return that session's halg
+ *  3. If policy-session, then return policy-session's halg
+ *
+ * Otherwise
+ *  return SHA256
```

Comment:
Whats the otherwise case(s)?

Reply:
Basically, password sessions and some commands like getrandom that don't require an auth-session at all.